### ISSUE TITLE
print-schema fails to generate schema for table with foreign key

### ISSUE THREAD

**Issue by Munksgaard** [diesel-rs/diesel#1948]
## Setup

```sql
BEGIN;

CREATE SCHEMA diesel_error

CREATE TABLE diesel_error.foo (
  id SERIAL PRIMARY KEY,
  date DATE NOT NULL
);

CREATE UNIQUE INDEX foo_id_date ON diesel_error.foo (id, date);

CREATE TABLE diesel_error.bar (
  id SERIAL PRIMARY KEY,
  foo_id INTEGER NOT NULL REFERENCES diesel_error.foo(id),
  date DATE,
  FOREIGN KEY (foo_id, date) REFERENCES diesel_error.foo(id, date)
);

COMMIT;
```

### Versions

- **Rust:** 1.32.0-nightly
- **Diesel:** 1.3.1
- **Database:** PostgreSQL 11.1
- **Operating System** Linux turing 4.4.0-141-generic #167-Ubuntu SMP Wed Dec 5 10:40:15 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

### Feature Flags

- **diesel:** postgres

## Problem Description

When trying to run `diesel print-schema` for the above schema, diesel prints the following error:

```
Unexpected null for non-null column
```

### What are you trying to accomplish?

I want to use an index that contains a nullable field. I realize that columns in which said field are `null` will not be part of the index, but that is acceptable to me.

### What is the expected output?

A working schema.rs file.

### What is the actual output?

Nothing

### Are you seeing any additional errors?

No.

### Steps to reproduce

Deploy the schema from above, and run `diesel print-schema` on it.

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) for similar issues.
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)

This has nothing to do with rust.

**Comment by Munksgaard** [comment]
In fact, it seems that I don't even need `bar.date` to be non-null. `diesel print-schema` fails on this schema as well:

```sql
BEGIN;

CREATE SCHEMA diesel_error

CREATE TABLE diesel_error.foo (
  id SERIAL PRIMARY KEY,
  date DATE NOT NULL
);

CREATE UNIQUE INDEX foo_id_date ON diesel_error.foo (id, date);

CREATE TABLE diesel_error.bar (
  id SERIAL PRIMARY KEY,
  foo_id INTEGER NOT NULL REFERENCES diesel_error.foo(id),
  date DATE NOT NULL,
  FOREIGN KEY (foo_id, date) REFERENCES diesel_error.foo(id, date)
);

COMMIT;
```

**Comment by Munksgaard** [comment]
For reference, the following schema works:

```sql
BEGIN;

CREATE SCHEMA diesel_error

CREATE TABLE diesel_error.foo (
  id SERIAL PRIMARY KEY,
  date DATE NOT NULL
);

CREATE UNIQUE INDEX foo_id_date ON diesel_error.foo (id, date);

CREATE TABLE diesel_error.bar (
  id SERIAL PRIMARY KEY,
  foo_id INTEGER NOT NULL REFERENCES diesel_error.foo(id),
  date DATE NOT NULL
);

COMMIT;
```

and results in this output:

```rust
pub mod diesel_error {
    table! {
        diesel_error.bar (id) {
            id -> Int4,
            foo_id -> Int4,
            date -> Date,
        }
    }

    table! {
        diesel_error.foo (id) {
            id -> Int4,
            date -> Date,
        }
    }

    joinable!(bar -> foo (foo_id));

    allow_tables_to_appear_in_same_query!(
        bar,
        foo,
    );
}
```

**Comment by weiznich** [comment]
I think composite foreign keys are currently not supported (at least in combination with `diesel print-schema`).
You may be able to workaround this by setting `diesel print-schema` up in such a way that it ignores those tables. Than you need to write the table definitions by yourself. The associations api will probably not be usable to. Also it should not be possible to use the `joinable!` macro to simplify join statements, but it should be possible to write queries involving a join over both tables by using explicit on clauses.

**Comment by sgrif** [comment]
Composite foreign keys are not supported, but we should be ignoring them, not erroring out.

**Comment by callym** [comment]
I've ran into this issue recently and trying to ignore the table in diesel.toml doesn't seem to work - if anyone has any pointers as to where to look to find where this error is coming from and how to fix it, I'm happy to have a go!

**Comment by sgrif** [comment]
@callym Somewhere in https://github.com/diesel-rs/diesel/blob/bfdddefaa80b732eb656e1ac01c92f10653c904f/diesel_cli/src/infer_schema_internals/information_schema.rs#L62-L116, a column should be nullable that isn't. As for which column, I'm not sure.

**Comment by robertohuertasm** [comment]
I'm new to Diesel so maybe I'm missing something. I was testing it to use it in a new project but I've just found this issue. It seems that having a foreign key to a non primary key is not supported. Is this a known limitation?

```sql
CREATE TABLE foo (
  id SERIAL PRIMARY KEY,
  name VARCHAR NOT NULL
);

CREATE UNIQUE INDEX if NOT EXISTS index_projects_name ON foo(name);

CREATE TABLE bar (
  id SERIAL PRIMARY KEY,
  parent VARCHAR NOT NULL REFERENCES foo(name)
);
```
When running the migration I get: `Unexpected null for non-null column`.

**Comment by weiznich** [comment]
@robertohuertasm Yes it's a known issue with the latest released version

Closed because already fixed by #2073 on master, which means the next release will contain this fix.

**Comment by robertohuertasm** [comment]
@weiznich thanks for your quick answer but just to be clear. Both issues are already fixed by #2073? Or are you referring only to compound fk?

**Comment by weiznich** [comment]
@robertohuertasm As far as I can tell the underlying issue is the same. Otherwise please open a new issue, as this is then clearly off topic here.
